python/jupyterlab-chat/src/chat-commands/providers/user-mention.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) Jupyter Development Team.
+ * Distributed under the terms of the Modified BSD License.
+ */
+
+import {
+  IChatCommandProvider,
+  IChatCommandRegistry,
+  ChatCommand,
+  IInputModel,
+  Avatar,
+  IUser
+} from '@jupyter/chat';
+import { JupyterFrontEndPlugin } from '@jupyterlab/application';
+import { User } from '@jupyterlab/services';
+import { JSONObject } from '@lumino/coreutils';
+import { LabChatModel } from 'jupyterlab-chat';
+
+export const mentionCommandsPlugin: JupyterFrontEndPlugin<void> = {
+  id: 'jupyterlab-chat-extension:mentionCommandsPlugin',
+  description: 'Plugin which adds user mention commands.',
+  autoStart: true,
+  requires: [IChatCommandRegistry],
+  activate: (app, registry: IChatCommandRegistry) => {
+    registry.addProvider(new MentionCommandProvider());
+  }
+};
+
+class MentionCommandProvider implements IChatCommandProvider {
+  public id: string = 'jupyter-chat:mention-commands';
+
+  // regex used to test the current word
+  private _regex: RegExp = /^@[\w-]*:?/;
+
+  async getChatCommands(inputModel: IInputModel, chatModel: LabChatModel) {
```

Comment:
Passing the entire `chatModel` to each command provider may be providing an excessive amount of scope. Ideally, chat command providers should only have control over the input, and should not be able to affect the entire chat state.

It seems like it may make more sense to replace the `chatModel: ChatModel` argument with a `chatContext: ChatContext` object, a custom object that provides read-only APIs to the chat, like `getUsers()`. That would also greatly simplify this method, since `MentionCommandProvider` could just call `chatContext.getUsers()` to get the list of users.

Reply:
:+1: I agree with you, and maybe that read-only chat model could be passed to the `ChatInput`, that way it could also be used for the buttons in the toolbar registry.

Reply:
@brichet That sounds good! Do you think that this should be done in a follow-up PR, or should we address it here?

Either way sounds fine to me, but I think we should solve this before the next release. Otherwise we might have to migrate Jupyter AI v3 to use `chatModel`, only to change it later to use `chatContext`.

Reply:
I'll update this PR